### PR TITLE
fix: fall back to CPU if GPU is not available

### LINKED ISSUE #1516: Test segfaults when no GPU can be found
### Description

Running `RUST_LOG=trace FIL_PROOFS_USE_GPU_TREE_BUILDER=1 cargo test --test stacked_vanilla test_stacked_porep_prove_verify` on a machine that doesn't have an Nvidia or AMD GPU segfaults.

### Acceptance criteria

It returns an error instead of a segfault.

### PR DESCRIPTION
**Pull request by vmx**
The current code expected that if the GPU-based tree building is enabled,
that there also is a working GPU available. When it is not, it panics.
This commit adds a fallback to CPU in case selecting a working GPU fails.

It can be verified by running

    FIL_PROOFS_USE_GPU_TREE_BUILDER=1 cargo test --test stacked_vanilla test_stacked_porep_prove_verify

if the machine has no working Nvidia or AMD GPU.

Fixes #1516.

### CODE REVIEW

**Review by cryptonemo — DISMISSED** [review]

File: storage-proofs-porep/src/stacked/vanilla/proof.rs [hunk 1]
```diff
@@ -594,14 +594,20 @@ impl<'a, Tree: 'static + MerkleTreeTrait, G: 'static + Hasher> StackedDrg<'a, Tr
...
-                    let tree_batcher = Some(
-                        Batcher::pick_gpu(max_gpu_tree_batch_size)
-                            .expect("failed to create tree gpu batcher"),
-                    );
-                    let column_batcher = Some(
-                        Batcher::pick_gpu(max_gpu_column_batch_size)
-                            .expect("failed to create col gpu batcher"),
-                    );
+                    let tree_batcher = match Batcher::pick_gpu(max_gpu_tree_batch_size) {
+                        Ok(b) => Some(b),
```
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply storage-proofs-porep/src/stacked/vanilla/proof.rs:598]: you are not printing the actual error

File: storage-proofs-porep/src/stacked/vanilla/proof.rs [hunk 2]
```diff
@@ -1005,11 +1014,7 @@ impl<'a, Tree: 'static + MerkleTreeTrait, G: 'static + Hasher> StackedDrg<'a, Tr
                         };
 
                         // If we get here, this is a final leaf batch: build a sub-tree.
-                        info!(
-                            "building base tree_r_last with GPU {}/{}",
-                            i + 1,
-                            tree_count
-                        );
+                        info!("building base tree_r_last {}/{}", i + 1, tree_count);
```
[the remark this answers is not in the document]
Comment (vmx) [reply storage-proofs-porep/src/stacked/vanilla/proof.rs:1017]: If it says "with GPU", I'd consider "working" being that it actually runs on the GPU. That may no longer be the case.

I propose logging "GPU" or "CPU" depending on whether it's running on the CPU or the GPU, so if people grep for it, they immediately see "oh it's running on the CPU, I need to check the logs why" and then see the previous warning.

[the remark this answers is not in the document]
Comment (cryptonemo) [reply storage-proofs-porep/src/stacked/vanilla/proof.rs:1017]: I disagree, it's important to tell which method is run (i.e. the codepath attempted).  It doesn't matter if it later fails (or rather, that's something else to look for in locks), which already can happen with lock issues, etc.

[the remark this answers is not in the document]
Comment (cryptonemo) [reply storage-proofs-porep/src/stacked/vanilla/proof.rs:1017]: I'm requesting we add it back -- I have specifically dug through logs with miners and we use that as a starting point for debugging.

**Pull request merged**